src/libcore/iter/iterator.rs

```diff
@@ -1403,14 +1473,10 @@ pub trait Iterator {
     /// ```
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn fold<B, F>(self, init: B, mut f: F) -> B where
+    fn fold<B, F>(mut self, init: B, mut f: F) -> B where
         Self: Sized, F: FnMut(B, Self::Item) -> B,
     {
-        let mut accum = init;
-        for x in self {
-            accum = f(accum, x);
-        }
-        accum
+        self.try_fold(init, move |acc, x| AlwaysOk(f(acc, x))).0
```

Comment:
I'm fairly new to rust and I know other languages implements their iterator methods based on fold. I imagine the original implementors of these methods know that also and if they use simple for loops is because an operation inside a for is very compiler friendly?
Creating the closure here + `if let`, etc. In `try_fold`. Unless the compiler is really really good it will create slower code, no?
 Have you tried iterating on more complex data than just integers?

Reply:
I think it's more that a `for` loop was the obvious way to do it, and the emphasis on internal iteration is a more recent idea.  For simple iterators, it should be a wash, but iterators like `Chain` can lift their conditionals out in a `fold` or `try_fold`, better than repeated `next` calls.

Reply:
@JordiPolo Here's a link to explore how this gets translated: https://godbolt.org/g/3ehFBV

There are a few things interacting here to make it relatively straight-forward for the compiler to turn this into good code.  Note the definition of the `AlwaysOk` type:
```rust
struct AlwaysOk<T>(pub T);
```
That means that wrapping something in `AlwaysOk` is actually not doing anything -- the memory layout doesn't change at all.  (Asterisk for potential ABI implications and that repr(rust) layout is subject to change, but that shouldn't be relevant in this case.)  Similarly, the `.0` at the end is also a type-level-only thing, as it doesn't need to change the representation at all.

The other thing that the compiler needs to be able to do is to know that the `?` operators in the try_fold materialization will never return early.  But see the `Try` impl:
```rust
impl<T> Try for AlwaysOk<T> {
    type Ok = T;
    type Error = !;
    ...
}
```
The "never type" `!` there is the canonical uninhabited type.  (There are others, like if you define `enum NoVariants {}`.)  Because uninhabited types have no valid values, it knows that an error can never happen, so it can completely remove the early-return paths, making it equivalent to normal fold.

Reply:
@JordiPolo You're right, it's asking the compiler to inline a lot, but it's not a miraculous effort, since closures like other things in Rust default to being *unboxed*.

Reply:
Thanks so much for the details, I think `!` is the magic I needed to understand, now I see the logic